src/lib/homestore_backend/index_kv.cpp

```diff
@@ -59,9 +59,98 @@ BlobManager::NullResult HSHomeObject::add_to_index_table(shared< BlobIndexTable
     return folly::Unit();
 }
 
+BlobManager::Result< homestore::MultiBlkId > HSHomeObject::move_to_tombstone(shared< BlobIndexTable > index_table,
+                                                                             const BlobInfo& blob_info) {
+    // since there may be a crash at any step of this function, we need to make sure the index table is always
+    // correct when we recover and reapply the raft log.
+    // the sequence here must be get, put and remove, so that we can keep the correctness of the index table.
+
+    BlobRouteKey index_key{BlobRoute{blob_info.shard_id, BlobRoute::KeyType::kTypeNormal, blob_info.blob_id}};
+    BlobRouteValue index_value;
+    homestore::BtreeSingleGetRequest get_req{&index_key, &index_value};
+    auto status = index_table->get(get_req);
+    if (status != homestore::btree_status_t::success) {
+        LOGERROR("Failed to get from index table {}", index_key.to_string());
+        return folly::makeUnexpected(BlobError::UNKNOWN_BLOB);
+    }
+
+    index_key.change_to_tombstone();
+    // we set a tombstone in key, of which GC can take the benifit to quickly finding the delete blkIDs by using
+    // the range query of key
+    homestore::BtreeSinglePutRequest put_req{&index_key, &index_value,
```

Comment:
As we already called blkallocator to free the PBAs,  technically the PBA is not belongs to HomeObj anymore.  IMO no point to store the PBA in index_value of tombstone,   the raft LSN makes more sense by providing more info.

Reply:
now, when we free pbas,  appendBlockAllocator only update `m_freeable_nblks` (sometimes , `m_last_append_offset` also), it does not store which pbas are freeed, so we have to store the pbas of tombstone for GC thread to determine which pbas should not be copied to a new chunk.
well, if appendBlockAllocator stores those freeed pbas some day, we can simplifiy the code here.

do not quite understand what infomation can LSN provide. can you explain more deeper?

Reply:
in current GC design flow,  when key is marked with tombstone, the value is no used as we will only filter-out the alive index key/value from the index table and only move that parts of data. we do not need to know the pba of one key with tombstone as will not touch that data anymore.

Reply:
no need to store the deleted pba, GC won't consume that pba, but only the state of that blob, if it was tomestoned, skip.